### DISCUSSION TITLE
Long form composition in Strudel

### DISCUSSION THREAD

**Discussion by jarmitage** [tidalcycles/strudel#629] [Ideas]
Tidal ref: https://tidalcycles.org/docs/reference/composition

@felixroos I'm guessing you've already thought a lot about how you would like to explore long(er) form composition in Strudel once other features are ticked off, care to share?

I've just been (re)prototyping long form composition in Tidal and almost reaching a good place now. I can see that with a bit more work, some redesigns of `ur` and `seqP` would essentially allow something like:

```haskell
patterns = [("percussion-a", s "bd"), ("percussion-b", s "bd*4"), ... ]
sequences = [("percussion", "a b"), ("bass", "~ a [~ b]"), ... ]
effects = [("reverb", # rvb "0.3:0.3"), ... ]

d1 $ sequences patterns effects
```

So similar to `ur`, but more comprehensive for capturing a the structure of an entire piece. You could then essentially playback a Tidal piece but then also alter it on the fly...

Tidal musicians / live coders in general are often performing multiple pieces sequentially and potentially mixing / remixing between them, so then there's another question of how this is facilitated by the composition strategy.

The advantage of `ur` is not just that the patterns become patternable, but also existing within a Map means they can potentially be referenced in other ways.

So in Strudel for example, would it make sense to explore defining patterns in JSON objects, and the referencing their keys in a pattern of patterns? JSONificiation might be best handled via parsing / UI / structure editor rather than literal JSON syntax in the REPL. If Strudel ends up offering block-based evaluation, I guess those blocks could be automatically referenced?

Bit of a ramble here, but overall it seems that a composition can be thought of as:
- Patterns
- Patterns of patterns
- Sequences of patterns of patterns
- Mechanisms for playing back the overall sequence while still being able to edit all the way down on the fly

Then composability between those layers (e.g. in my example ) would be the "Tidaliest" way to do long form composition. So similar to how Strudel has drum samples standardised around "bd sn" etc, you might want the same for composition names and their subsections (e.g. "a b c").

**Comment by felixroos** [comment]
there is the arrange function, example: https://strudel.tidalcycles.org/?THqzJZBxTDSS doc: https://strudel.tidalcycles.org/learn/factories#arrange

**Reply by jarmitage** [reply]
Yeah this is great, similar to `timeCat` essentially? https://tidalcycles.org/docs/reference/concatenation/#timecat

**Reply by felixroos** [reply]
same syntax, but works a little bit differently to make sure nothing is shuffled / speed up. it also uses timeCat under the hood: https://github.com/tidalcycles/strudel/blob/90a58858ca6ad685a12c9b1de1df4ff7331dc2e8/packages/core/pattern.mjs#L1297 

if you replace arrange with timeCat in the example, weird things happen: https://strudel.tidalcycles.org/?frFP-8QUXtCL
maybe timeCat in tidal works like arrange in strudel, not sure.. if so, then timeCat and arrange should be synonyms in strudel i guess. can you try it out?

**Reply by felixroos** [reply]
the weird stuff only happens if you timeCat things that span multiple cycles, and maybe they also need different lengths

**Reply by jarmitage** [reply]
Yeah I think it's similar for any concatenation / composition stuff in Tidal. The weird stuff can be quite cool but also confusing, I imagine Strudel's highlighter might be very handy in this regard!

**Reply by felixroos** [reply]
yep but if you squeeze all items of the arrangement into one cycle, then timeCat them and slow them down again, it works (as arrange does)

**Comment by jarmitage** [comment]
@yaxu do Tidal 2 and sequences change things in terms of how you would (re)approach things like `ur` / "higher level patterns"?

**Comment by felixroos** [comment]
kind of related function: throw in a list of patterns + playtime + fadetime to get automated dj-style mixing between them: https://strudel.tidalcycles.org/?YtkX3prz4D7h though this approach only works when each "track" is a single pattern

**Comment by felixroos** [comment]
related: https://github.com/tidalcycles/strudel/issues/670

**Reply by gogins** [reply]
Is there any way to use a slider to turn a pattern on and off? I would like something like `catIf` in the use case below. Suppose I have:
```
let A = "a b c";
let B = "d e"; 
let C = "f g";
A().catIf(B, slider(1, 0, 1, 1).C().piano()
```
So, if the slider's value is 1, the sequence of Patterns is A, B, C. But, if the slider's value is 0, the sequence of Patterns is A, C, with no gap between A and C.

I looked at `arrange` because I thought that might do the job, but I found it impossible to use `slider` to set the number of repetitions, and I also got a divide by 0 error if I set the number of repetitions to 0.
